### ISSUE TITLE
[FEA] Faster dataframe to cupy conversion when dataframe is a single allocation

### ISSUE THREAD

**Issue by beckernick** [rapidsai/cudf#12928]
When we convert a dataframe to a cupy array, we iterate over each column (as they’re independent allocations) and assign each one to a column in an empty matrix. This means it can be slow for thousands or millions of small columns.

In a select set of circumstances, all of the columns in a DataFrame may be part of a single, contiguous allocation of memory. One scenario in which this can occur is after a call to transpose. It would be nice if, in this scenario, we didn't need to iterate over every column when converting to a cupy array.

A real-world example of when this can matter is if a user is trying to run a dot product after a calling transpose. Because of the bottleneck, we're slower than pandas by quite a bit.

```python
import cudf
import cupy as cp
import pandas as pd

nrows = 10000
ncols = 4

gdf = cudf.DataFrame(cp.random.randint(0, 1000, size=(nrows, ncols))/10000)
pdf = gdf.to_pandas()

%time gdf.T.dot(gdf)
%time pdf.T.dot(pdf)
CPU times: user 1.52 s, sys: 3.96 ms, total: 1.53 s
Wall time: 1.53 s
CPU times: user 912 µs, sys: 41 µs, total: 953 µs
Wall time: 855 µs
```

If we were to do any special casing here, we'd want to closely evaluate any impact on performance for the more general case, as the dataframe to cupy codepath is used across the board.

**Comment by shwina** [comment]
The resulting CuPy array would need to be F-contiguous, IIUC. Would that be alright?

**Comment by wence-** [comment]
> The resulting CuPy array would need to be F-contiguous, IIUC. Would that be alright?

Yes, since cupy dot handles that fine.

The right answer here (since the `.T` is already expensive) is to do:
```
tmp = gdf.values
result = cudf.DataFrame(tmp.T.dot(tmp))
```
But that is certainly not as seamless

**Comment by beckernick** [comment]
Agreed, F contiguousness would be fine. With that said, if we decide that a fastpath for this sceniaro has a non-trivial cost (in performance or complexity) associated with it and isn't worth the cost, that's also reasonable.

For completeness, when the user hit this originally they switched to CuPy to do the operations (like Lawrence suggested).

**Comment by vyasr** [comment]
Related: #11648
